source/fuzz/replayer.h

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SOURCE_FUZZ_REPLAYER_H_
+#define SOURCE_FUZZ_REPLAYER_H_
+
+#include <memory>
+#include <vector>
+
+#include "source/fuzz/protobufs/spirvfuzz_protobufs.h"
+#include "spirv-tools/libspirv.hpp"
+
+namespace spvtools {
+namespace fuzz {
+
+// Transforms a SPIR-V module into a semantically equivalent SPIR-V module by
+// applying a series of pre-defined transformations.
+class Replayer {
+ public:
+  // Possible statuses that can result from running the replayer.
+  enum ReplayerResultStatus {
+    kComplete,
+    kFailedToCreateSpirvToolsInterface,
+    kInitialBinaryInvalid,
+    kInitialFactsInvalid,
+  };
+
+  // Constructs a replayer from the given target environment.
+  explicit Replayer(spv_target_env env);
+
+  // Disables copy/move constructor/assignment operations.
+  Replayer(const Replayer&) = delete;
+  Replayer(Replayer&&) = delete;
+  Replayer& operator=(const Replayer&) = delete;
+  Replayer& operator=(Replayer&&) = delete;
+
+  ~Replayer();
```

Comment:
Is this necessary? The implementation is just the default. Probably best not to include unless it's different.

Reply:
It does seem to be necessary, yes - if I remove the destructor then I get compile errors due to the use of unique_ptr in the class.  I confess that I do not fully understand the reason for these errors, but I believe explicitly declaring a default destructor really is required here.

Reply:
Can you move the implementation into the header in that case? Good to know it's not doing anything complicated.

Reply:
Done.

Reply:
Actually moving the implementation into the header doesn't work.  This is to do with the use of std::unique_ptr and forward declaration.  It seems that by being inline the destructor gets deleted.

I notice that opt/optimizer.hpp and opt/optimizer.cpp suffer from a similar issue - if the default destructor for Optimizer is moved in to the .hpp file, tools/opt.cpp fails to compile.